docs(serverless): secure your function/container (#1258)
scaleway/docs-content: serverless/containers/how-to/secure-a-container.mdx

---
meta:
  title: How to secure a container
  description: This page explains how to secure a Serverless container
content:
  h1: How to secure a container
  paragraph: This page explains how to  how to secure a Serverless container
tags: containers secure private
dates:
  validation: 2023-02-01
  posted: 2023-02-01
categories:
  - serverless
---

This page explains how to secure your container.

<Macro id="iam-requirements" />

<Message type="requirement">
  - You have an account and are logged into the [Scaleway console](https://console.scaleway.com)
  - You have [created a containers namespace](/serverless/containers/how-to/create-a-containers-namespace/)
  - You have [deployed a container](/serverless/containers/how-to/deploy-a-container/)
</Message>

## Use Secrets to store access keys and sensible information

> Instead of using environment variables (which are stored in clear) for your containers, use Secrets. These are information usable through environment variables but are stored encrypted. 

Reviewer: ```suggestion
Instead of using environment variables (which are stored in clear text) for your containers, use Secrets. These are pieces of information that can be used via environment variables, but are encrypted in storage. 
```

### Configure Secrets from the Scaleway Console

1. Click **Containers** in the **Serverless** section of the side menu. The containers page displays.

> 2. Click the relevant containers namespace. 

Reviewer: ```suggestion
2. Click the relevant container namespace. 
```

3. Click the name of the container for which you want to define Secrets.
4. Click the **Deployment** tab.
5. Scroll to the **Secrets** section of the page and click **Add a new secret**. Enter the key and value for your Secret. Repeat for additonal secrets.
    <Message type="important">
        Be careful when you type your Secrets. You will not be able to read the value in the console once submitted.
    </Message>
6. Click **Deploy a Container** to submit your Secrets and redeploy your container. 

### Configure Secrets using the Serverless framework

> Add `secret` to your containers description (more information in the [plugin documentation](https://github.com/scaleway/serverless-scaleway-containers#security-and-secret-management))

Reviewer: ```suggestion
Add `secret` to your containers description (more information in the [plugin documentation](https://github.com/scaleway/serverless-scaleway-containers#security-and-secret-management)).
```

> Add `secret` to your containers description (more information in the [plugin documentation](https://github.com/scaleway/serverless-scaleway-containers#security-and-secret-management)).

Reviewer 2: link is dead, replace with : https://github.com/scaleway/serverless-scaleway-functions#security-and-secret-management

Author: ```suggestion
Add `secret` to your containers description (more information in the [plugin documentation](https://github.com/scaleway/serverless-scaleway-functions#security-and-secret-management)).
```

> We recommend to use them with global environment variables or a `.env` file stored independently (and kept secret).

Reviewer: ```suggestion
We recommend using them with global environment variables or a `.env` file stored independently (and kept secret).
```

``` 
secret:
      secret_1: ${env:SCW_SECRET_KEY} 
```
<Message type="important">

>     We recommend you to not commit in a Version Control System (VCS), and to not share your Project ID or access key to ensure the security of your configuration file, which may contain sensitive data.

Reviewer: ```suggestion
    We strongly suggest that you do not commit this in a Version Control System (VCS), and do not share your Project ID or access key. This helps to ensure the security of your configuration file, which may contain sensitive data.
```

>     We strongly suggest that you do not commit this in a Version Control System (VCS), and do not share your Project ID or access key. This helps to ensure the security of your configuration file, which may contain sensitive data.

Reviewer 2: Project ID or access key ->  Project ID nor access key ? 🤔

</Message>

### Configure Secrets using Terraform

Add the following [resource description](https://registry.terraform.io/providers/scaleway/scaleway/latest/docs/resources/container) in Terraform:

```
secret_environment_variables = { "key" = "secret" }
```

## Restrict access to your containers

You can set Serverless containers as private if you want, to protect them from unwanted or unauthorized calls. 
Unauthenticated calls will be rejected, and your container will not be triggered. This feature is handy if an event triggers your container (CRON, and soon SQS or NATS trigger) or if you put them behind an API gateway or a proxy server (see examples in serverless-examples).

### Restrict access from the Scaleway Console

> 1. Click **containers** in the **Serverless** section of the side menu. The containers page displays.

Reviewer: ```suggestion
1. Click **Containers** in the **Serverless** section of the side menu. The containers page displays.
```

> 2. Click the relevant containers namespace. 

Reviewer: ```suggestion
2. Click the relevant container namespace. 
```

3. Click the name of the container for which you want to define Secrets.
4. Click the **Security** tab.
5. Set the **Privacy Policy** of the container to **Private**. 

> 6. If required create an access token for your container:

Reviewer: ```suggestion
6. If required, create an access token for your container:
```

        </Message>

### Restrict access using the Serverless framework

Set `privacy: private` in your container's description.

You can generate access tokens for your container using the Scaleway console: 

1. Click **Containers** in the **Serverless** section of the side menu. The containers page displays.
2. Click the relevant container namespace. 
3. Click the name of the container for which you want to define Secrets.
4. Click the **Security** tab.
5. Click **Generate token** in the **Tokens** section of the page. A pop-up displays.
6. Enter a description and set an optional expiration date. To generate a permanently-valid token, tick **No expiration date**. Then click **Generate token**.
    The token is displayed. Copy and store the token in a secure environment.
    <Message type="important">
        The token key will be shown only once. Copy and safely store it before leaving this page.
    </Message>

### Restrict access using Terraform

Set `privacy = "private"` in your Terraform [resource description](https://registry.terraform.io/providers/scaleway/scaleway/latest/docs/resources/container).

> You can generate access credentials to inject in other applications (containers, containers ...) directly from Terraform using the [container_token resource](https://registry.terraform.io/providers/scaleway/scaleway/latest/docs/resources/container_token)

Author: ```suggestion
You can generate access credentials to inject in other applications (containers, functions, ...) directly from Terraform using the [container_token resource](https://registry.terraform.io/providers/scaleway/scaleway/latest/docs/resources/container_token)
```
